src/axom/core/utilities/BitUtilities.hpp

```diff
@@ -18,11 +18,12 @@
 #include "axom/core/Macros.hpp"
 #include "axom/core/Types.hpp"
 
+// CUDA intrinsics: https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__INT.html
+// HIP intrinsics: https://rocm.docs.amd.com/projects/HIP/en/latest/reference/kernel_language.html
+
 // Check for and setup defines for platform-specific intrinsics
 // Note: `__GNUC__` is defined for the gnu, clang and intel compilers
-#if defined(__CUDACC__)
```

Comment:
There were specific cases where the cuda compiler was falling into cases where it should not have. If I remember correctly it was CUDA + avx512, which, to my knowledge, do not have any pipelines to guard against.

Here is the issue here:

https://github.com/LLNL/axom/issues/798

and the PR:

https://github.com/LLNL/axom/pull/1088

Reply:
In public headers, you have to be careful about how you present define checks that get evaluated in downstream libs.

Best practice is have your own defines that reflect the choices your lib baked in during its compile -- and use those instead of rely on compiler specific defines that may change downstream. 

Otherwise you end up bad places, ODRs are one such place. 

Defines related to cuda, hip, and openmp are all ripe for this kind of excitement.

If we do need a specific case for CUDA, then we should use `AXOM_USE_CUDA`  in that place ( not `__CUDACC__ ` & not `AXOM_USE_GPU`, unless same problem exists for hip)

Reply:
So do we want something more like the following?

```
#if defined(_WIN64) && (_MSC_VER >= 1600)
  #define _AXOM_CORE_USE_INTRINSICS_MSVC
  #include <intrin.h>
#elif defined(__x86_64__) && defined(__GNUC__)
  #define _AXOM_CORE_USE_INTRINSICS_GCC
  #if !defined(__CUDACC__)
    #include <x86intrin.h>
  #endif
#elif defined(__powerpc64__) && (defined(__GNUC__) || defined(__ibmxl__))
  #define _AXOM_CORE_USE_INTRINSICS_PPC
#endif
```

Reply:
Or perhaps the better approach is to move what we can into a source file instead of having it here in a header.

Reply:
Sorry, I missed Cyrus's comment. I changed the code to use AXOM_USE_CUDA.

Reply:
This still prevents intrinsics from being used in the host side functions, but at least there are no ODR violations.

Reply:
I'm still wondering about putting the definitions of these functions in a source file and letting the compiler take care of inlining.